### PR TITLE
ControlledToken array 

### PR DESCRIPTION
**Pull request by aodhgan**
as per https://linear.app/pooltogether/issue/POOL-1341/refactor-controlled-tokens-in-prize-pool-to-use-array-instead-of

### CODE REVIEW

File: contracts/prize-pool/PrizePool.sol [hunk 1]
```diff
@@ -190,7 +190,7 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
     initializer
   {
     require(address(_reserveRegistry) != address(0), "PrizePool/reserveRegistry-not-zero");
-    _tokens.initialize();
+
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:193]: Let's initialize the array here with the correct length

File: contracts/prize-pool/PrizePool.sol [hunk 2]
```diff
@@ -781,8 +781,8 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
   /// @param _controlledToken The controlled token to add.  Cannot be a duplicate.
   function _addControlledToken(ControlledTokenInterface _controlledToken) internal {
     require(_controlledToken.controller() == this, "PrizePool/token-ctrlr-mismatch");
-    _tokens.addAddress(address(_controlledToken));
-
+    
+    _tokens.push(_controlledToken);
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:785]: here let's instead pass an index and set the token at that index (save 5k gas per token)

File: contracts/prize-pool/PrizePool.sol [hunk 3]
```diff
@@ -842,12 +842,12 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
...
   function _tokenTotalSupply() internal view returns (uint256) {
     uint256 total = reserveTotalSupply;
-    address currentToken = _tokens.start();
-    address tokenEnd = _tokens.end();
-    while (currentToken != tokenEnd) {
-      total = total.add(IERC20Upgradeable(currentToken).totalSupply());
-      currentToken = _tokens.next(currentToken);
+    ControlledTokenInterface[] memory tokens = _tokens; // SLOAD  
+    
+    for(uint256 i = 0; i < tokens.length; i++){
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:847]: let's pull the length into a var to save gas

File: contracts/prize-pool/PrizePool.sol [hunk 4]
```diff
@@ -862,8 +862,12 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
   /// @dev Checks if a specific token is controlled by the Prize Pool
   /// @param controlledToken The address of the token to check
   /// @return True if the token is a controlled token, false otherwise
-  function _isControlled(address controlledToken) internal view returns (bool) {
-    return _tokens.contains(controlledToken);
+  function _isControlled(ControlledTokenInterface controlledToken) internal view returns (bool) {
+    ControlledTokenInterface[] memory tokens = _tokens; // SLOAD
+    for(uint256 i = 0; i < tokens.length; i++) {
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:867]: pull out token length as var to save gas

**Review by asselstine — CHANGES_REQUESTED** [review]

File: contracts/prize-pool/PrizePool.sol [hunk 5]
```diff
@@ -778,11 +779,13 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
...
   /// @notice Adds a new controlled token
-  /// @param _controlledToken The controlled token to add.  Cannot be a duplicate.
-  function _addControlledToken(ControlledTokenInterface _controlledToken) internal {
+  /// @param _controlledToken The controlled token to add.
+  /// @param index The index to add the controlledToken
+  function _addControlledToken(ControlledTokenInterface _controlledToken, uint256 index) internal {
     require(_controlledToken.controller() == this, "PrizePool/token-ctrlr-mismatch");
-    _tokens.addAddress(address(_controlledToken));
-
+    require(address(_controlledToken) != address(0), "PrizePool/controlledToken-not-zero");
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:786]: ✂️  

The previous line precludes the possilibity of it being zero!

File: contracts/prize-pool/PrizePool.sol [hunk 6]
```diff
@@ -862,8 +866,14 @@ abstract contract PrizePool is PrizePoolInterface, OwnableUpgradeable, Reentranc
   /// @dev Checks if a specific token is controlled by the Prize Pool
   /// @param controlledToken The address of the token to check
   /// @return True if the token is a controlled token, false otherwise
-  function _isControlled(address controlledToken) internal view returns (bool) {
-    return _tokens.contains(controlledToken);
+  function _isControlled(ControlledTokenInterface controlledToken) internal view returns (bool) {
```
Comment (asselstine) [inline contracts/prize-pool/PrizePool.sol:869]: Ah!  While you're here: let's add a public accessor!

**Review by asselstine — CHANGES_REQUESTED** [review]

File: contracts/prize-pool/PrizePool.sol (hunk 6 above)
Comment (aodhgan) [reply contracts/prize-pool/PrizePool.sol]: Sure!

**Review by asselstine — APPROVED** [review]

**Pull request merged**